### PR TITLE
Add initial openSUSE support

### PR DESCRIPTION
**Pull request by EXONER4TED**
Signed-off-by: Logan Bond <lbond@secureworks.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area cmd

/area pkg

> /area docs

> /area tests

**What this PR does / why we need it**:
Adds initial support for OpenSUSE 🙂 

**Which issue(s) this PR fixes**:

Fixes # N/A

**Special notes for your reviewer**:
I have a companion PR on kernel-crawler open to add support there too. Kernel-crawler's output helped build the URLs for this PR: https://github.com/falcosecurity/kernel-crawler/pull/59

**Does this PR introduce a user-facing change?**:
Adds initial OpenSUSE support 🙂 

```release-note
Add initial OpenSUSE support!
```

### CODE REVIEW

**Comment by EXONER4TED** [comment]
Here's some samples I ran through:
```
╰─❯ _output/bin/driverkit docker --output-module /tmp/test.ko --kernelrelease 5.3.18-lp152.19.2.x86_64 --target opensuse -l debug
...
...
...
DEBU  + modinfo /tmp/driver/module.ko      
DEBU �filename:       /tmp/driver/module.ko 
DEBU schema_version: 2.1.0                        
DEBU api_version:    2.0.0                        
DEBU build_commit:   master                       
DEBU version:        master                       
DEBU author:         the Falco authors            
DEBU license:        GPL                          
DEBU suserelease:    openSUSE Leap 15.2           
DEBU srcversion:     36C8B703CDEAE692FCFCEED      
DEBU depends:                                     
DEBU retpoline:      Y                            
DEBU name:           falco                        
DEBU vermagic:       5.3.18-lp152.19-default SMP mod_unload modversions  
DEBU parm:           g_buffer_bytes_dim:This is the dimension of a single per-CPU buffer in bytes. Please note: this buffer will be mapped twice in the process virtual memory, so pay attention to its size. 
DEBU parm:           max_consumers:Maximum number of consumers that can simultaneously open the devices (uint) 
DEBU parm:           verbose:Enable verbose logging (bool) 
DEBU log pipe close                                error=EOF
INFO kernel module available                       path=/tmp/test.ko
```

**Comment by EXONER4TED** [comment]
Fixing a few more resolving URLs, but then it should be good for review.

**Comment by EXONER4TED** [comment]
Fixed the remaining possible URLs, here are a couple more examples (with kernels 6.0+!):
```
╰─❯ _output/bin/driverkit docker --output-module /tmp/test.ko --kernelrelease 6.0.5-3.1.g7359656.x86_64 --target opensuse -l debug
...
...
...
DEBU  + modinfo /tmp/driver/module.ko      
DEBU �filename:       /tmp/driver/module.ko 
DEBU schema_version: 2.1.0                        
DEBU api_version:    2.0.0                        
DEBU build_commit:   master                       
DEBU version:        master                       
DEBU author:         the Falco authors            
DEBU license:        GPL                          
DEBU suserelease:    openSUSE Tumbleweed          
DEBU srcversion:     7056B12CCC80C546049066D      
DEBU depends:                                     
DEBU retpoline:      Y                            
DEBU name:           falco                        
DEBU vermagic:       6.0.5-3.g7359656-default SMP preempt mod_unload modversions  
DEBU parm:           g_buffer_bytes_dim:This is the dimension of a single per-CPU buffer in bytes. Please note: this buffer will be mapped twice in the process virtual memory, so pay attention to its size. 
DEBU parm:           max_consumers:Maximum number of consumers that can simultaneously open the devices (uint) 
DEBU parm:           verbose:Enable verbose logging (bool) 
DEBU log pipe close                                error=EOF
INFO kernel module available                       path=/tmp/test.ko
```
Example 2:
```
╰─❯ _output/bin/driverkit docker --output-module /tmp/test.ko --kernelrelease 6.0.3-1.1.x86_64 --target opensuse -l debug
...
...
...
DEBU  + modinfo /tmp/driver/module.ko      
DEBU �filename:       /tmp/driver/module.ko 
DEBU schema_version: 2.1.0                        
DEBU api_version:    2.0.0                        
DEBU build_commit:   master                       
DEBU version:        master                       
DEBU author:         the Falco authors            
DEBU license:        GPL                          
DEBU suserelease:    openSUSE Tumbleweed          
DEBU srcversion:     7056B12CCC80C546049066D      
DEBU depends:                                     
DEBU retpoline:      Y                            
DEBU name:           falco                        
DEBU vermagic:       6.0.3-1-default SMP preempt mod_unload modversions  
DEBU parm:           g_buffer_bytes_dim:This is the dimension of a single per-CPU buffer in bytes. Please note: this buffer will be mapped twice in the process virtual memory, so pay attention to its size. 
DEBU parm:           max_consumers:Maximum number of consumers that can simultaneously open the devices (uint) 
DEBU parm:           verbose:Enable verbose logging (bool) 
DEBU log pipe close                                error=EOF
INFO kernel module available                       path=/tmp/test.ko
```

File: pkg/driverbuilder/builder/opensuse.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+func (o *opensuse) Name() string {
+	return TargetTypeOpenSUSE.String()
+}
+
+func (o *opensuse) TemplateScript() string {
+	return opensuseTemplate
+}
+
+func (o *opensuse) URLs(_ Config, kr kernelrelease.KernelRelease) ([]string, error) {
```
Comment (FedeDP) [inline pkg/driverbuilder/builder/opensuse.go:65]: Neat! I'd let opensuse implement the MinimumURLs() interface. For an example, see Debian builder :)

Comment (EXONER4TED) [reply pkg/driverbuilder/builder/opensuse.go:74]: Done 🙂 I still left in the checks for the individual packages types. Even if 2 URLs are found, still need to make sure it's one of each appropriate type.

File: pkg/driverbuilder/builder/opensuse.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		fmt.Sprintf("kernel-devel-%s%s.rpm", kr.Fullversion, kr.FullExtraversion),
+		kr.Architecture.ToNonDeb(),
+		"noarch",
+	)
+
+	// get all possible URLs
+	possibleURLs := buildURLs(kr, kernelDefaultDevelPattern, kernelDevelNoArchPattern)
+
+	// trim the list to only resolving URLs
+	urls, err := getResolvingURLs(possibleURLs)
```
Comment (dwindsor) [inline pkg/driverbuilder/builder/opensuse.go:88]: Not necessarily for this PR, but do you think it's worthwhile to add a caching layer on top of `getResolvingUrls()`? For instance create a map `kernelrelease.KernelRelease[]string map` that would map a kernelrelease to the results of (expensive) calls to `getResolvingUrls()`. Users who run kernel-crawler/driverkit internally would benefit from not repeatedly issuing requests for URLs that will likely repeatedly fail.

**Review by dwindsor — APPROVED** [review]

File: pkg/driverbuilder/builder/opensuse.go (hunk 2 above)
Comment (EXONER4TED) [reply pkg/driverbuilder/builder/opensuse.go:88]: I had this same thought as well, and right now there's not really a good solution that I can think of... The problem with SUSE, at least as far as I can tell, is that they only host a single kernelrelease for a given distro release at a time. So you can't know beforehand where that kernel release is hosted on the myriad of sites (unless you have kernel crawler tell you, but that's still expensive to run). But when someone passes in `--kernelrelease 6.0.5-3.1.g7359656.x86_64`, there's no way to know (without a bunch of hardcoding) which openSUSE release that's for..

I think the ultimate solution is there needs to be a better way to discover the kernel package urls in general... Part of me wonders if the container-based approach and some way to query directly from the OS package manager might be better. 

But I knew this URL-based solution was faster to write and gets us a starting place. Another solution is to just remove the `getResolvingUrls()` entirely. The whole slew of possible URLs will be passed to the next bits of code, but 🤷 .

**Pull request merged**